bin/s2nc.c

```diff
@@ -186,223 +195,243 @@ int main(int argc, char *const *argv)
-        print_s2n_error("Error setting configuration");
-        exit(1);
-    }
+        if (s2n_connection_set_fd(conn, sockfd) < 0) {
+            print_s2n_error("Error setting file descriptor");
+            exit(1);
+        }
 
-    if (s2n_set_server_name(conn, server_name) < 0) {
-        print_s2n_error("Error setting server name");
-        exit(1);
-    }
+        /* Update session state in connection if exists */
+        if (session_state_length > 0 && s2n_connection_set_session(conn, session_state, session_state_length) < 0) {
+            print_s2n_error("Error setting session state in connection");
+            exit(1);
+        }
 
-    if (s2n_connection_set_fd(conn, sockfd) < 0) {
-        print_s2n_error("Error setting file descriptor");
-        exit(1);
-    }
+        /* See echo.c */
+        int ret = negotiate(conn);
 
-    /* See echo.c */
-    int ret = negotiate(conn);
+        if (ret != 0) {
+            /* Error is printed in negotiate */
+            return -1;
+        }
 
-    if (ret != 0) {
-        /* Error is printed in negotiate */
-        return -1;
-    }
+        printf("Connected to %s:%s\n", host, port);
 
-    printf("Connected to %s:%s\n", host, port);
+        /* Save session state from connection if reconnect is enabled */
+        if (reconnect > 0) {
+            if (!s2n_connection_get_session_id(conn)) {
+                printf("Endpoint sent empty session id so cannot resume session\n");
+                exit(1);
+            }
+            free(session_state);
+            session_state_length = s2n_connection_get_session_length(conn);
+            session_state = calloc(session_state_length, sizeof(uint8_t));
+            if (s2n_connection_get_session(conn, session_state, session_state_length) != session_state_length) {
+                print_s2n_error("Error getting serialized session state");
+                exit(1);
+            }
+        }
 
-    if (echo_input != 1) {
-        return 0;
-    }
+        if (echo_input == 1) {
+            echo(conn, sockfd);
+        }
 
-    echo(conn, sockfd);
+        s2n_blocked_status blocked;
+        s2n_shutdown(conn, &blocked);
 
-    s2n_blocked_status blocked;
-    if (s2n_shutdown(conn, &blocked) < 0) {
-        print_s2n_error("Error calling s2n_shutdown");
-        exit(1);
-    }
+        if (s2n_connection_free(conn) < 0) {
+            print_s2n_error("Error freeing connection");
+            exit(1);
+        }
 
-    close(sockfd);
+        if (s2n_config_free(config) < 0) {
+            print_s2n_error("Error freeing configuration");
+            exit(1);
+        }
 
-    if (s2n_connection_free(conn) < 0) {
-        print_s2n_error("Error freeing connection");
-        exit(1);
-    }
+        close(sockfd);
+        reconnect--;
 
-    if (s2n_config_free(config) < 0) {
-        print_s2n_error("Error freeing configuration");
-        exit(1);
-    }
+    } while (reconnect >= 0);
```

Comment:
The loop looks huge, should we reorganize it into functions?

Reply:
Moved most of the s2n_config related setup into a function, I did not move more because they need lost of parameters.